sql/analyzer/locks/lock_manager.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package locks
+
+import (
+	"sync"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+type ExclusiveLock struct {
+	clientId uint32
+	mu       *sync.Mutex
+}
+
+type LockManager interface {
+	// HoldTableLock informs the LockManager to reserve a table lock by the called client.
+	HoldTableLock(ctx *sql.Context, databaseName, tableName string) error
+
+	// PollTableLock polls the status of a table and exits when the table lock is not held by any other client except the
+	// current client.
+	PollTableLock(ctx *sql.Context, databaseName, tableName string) error
+
+	// ReleaseTableLocksHeldByClient release all held locks by the given client id.
+	ReleaseTableLocksHeldByClient(ctx *sql.Context, id uint32) error
+}
+
+type lockManagerImpl struct {
+	locksMap map[string]map[string]*ExclusiveLock // we will need to update this
+	catalog  sql.Catalog
+	mu       sync.RWMutex
+}
+
+var _ LockManager = &lockManagerImpl{}
+
+func NewLockManager(c sql.Catalog) *lockManagerImpl {
+	dbMap := make(map[string]map[string]*ExclusiveLock)
+
+	return &lockManagerImpl{locksMap: dbMap, catalog: c}
+}
+
+func (l *lockManagerImpl) HoldTableLock(ctx *sql.Context, databaseName, tableName string) error {
```

Comment:
what are the merits of locking within the manager vs having the manager hand out a common copy of the lock to all relevant consumers?